### ISSUE TITLE
关于可视化代码

### ISSUE THREAD

**Issue by saltever** [pangdatangtt/UniNet#9]
感谢您在 GitHub 上分享 [项目名称 / 代码仓库地址]，这对我在工业缺陷检测领域的学习和实践帮助很大！目前我正在使用您的代码进行相关实验，发现代码在缺陷检测的流程中运行正常，但缺少可视化部分，这让我在调试和理解检测结果时存在一定困难

**Comment by pangdatangtt** [comment]
感谢您对我们工作的关注。最近，我正忙于竞赛答辩和26 fall PhD面试的事情，所以不能及时回复问题。可视化代码你可以参考[RD](https://github.com/hq-deng/RD4AD/blob/main/test.py#L137)和[ReContrast](https://github.com/guojiajeremy/ReContrast/blob/master/utils.py#L359)方法中的代码，我们的方法是基于他们的代码上实现的。感谢您的关注！

**Comment by saltever** [comment]
十分感谢您的的解答，对我很有帮助，可视化已经融合了，祝你答辩顺利，fall phd 面试顺利，成功上岸！



---- 回复的原邮件 ----
| 发件人 | Shun ***@***.***> |
| 日期 | 2025年05月17日 17:41 |
| 收件人 | ***@***.***> |
| 抄送至 | ***@***.***>***@***.***> |
| 主题 | Re: [pangdatangtt/UniNet] 关于可视化代码 (Issue #9) |
pangdatangtt left a comment (pangdatangtt/UniNet#9)

感谢您对我们工作的关注。最近，我正忙于竞赛答辩和26 fall PhD面试的事情，所以不能及时回复问题。可视化代码你可以参考RD和ReContrast方法中的代码，我们的方法是基于他们的代码上实现的。感谢您的关注！

**Comment by xsjcrl** [comment]
您好，我不知道如何融合可视化代码，之前也没有做过这方面的工作，您可以分享下可视化代码吗？非常感谢！

**Comment by xsjcrl** [comment]
非常感谢您的回复，感谢您愿意分享可视化部分代码，想问下在哪里查看呀

**Comment by saltever** [comment]
> 非常感谢您的回复，感谢您愿意分享可视化部分代码，想问下在哪里查看呀
在utils.py，添加，在test.py调用就可以了

def visualize(model, dataloader, device, _class_='None', save_name='save'):
    model.eval()
    save_dir = os.path.join('./visualize', save_name)
    os.makedirs(save_dir, exist_ok=True)  
    with torch.no_grad():
        for idx, (sample, label, gt) in enumerate(dataloader):
            if isinstance(sample, (list, tuple)):  
                img = sample[0] 
            else:
                img = sample
            img = img.to(device)
            en, de = model(img)
            anomaly_map, _ = cal_anomaly_map(en, de, img.shape[-1], amap_mode='a')
            anomaly_map = gaussian_filter(anomaly_map, sigma=4)
            heatmap = min_max_norm(anomaly_map)
            heatmap = cvt2heatmap(heatmap * 255)
            img_np = img.permute(0, 2, 3, 1).cpu().numpy()[0]
            img_np = img_np * np.array([0.229, 0.224, 0.225]) + np.array([0.485, 0.456, 0.406])
            img_np = (img_np * 255).astype('uint8')
            hm_on_img = show_cam_on_image(img_np, heatmap)
            
            save_dir_class = os.path.join(save_dir, str(_class_))
            os.makedirs(save_dir_class, exist_ok=True)
            name = f"sample_{idx}"
            if hasattr(dataloader.dataset, 'samples'):  
                try:
                    img_path = dataloader.dataset.samples[idx][0]
                    dir_name = os.path.basename(os.path.dirname(img_path))
                    base_name = os.path.splitext(os.path.basename(img_path))[0]
                    name = f"{dir_name}_{base_name}"
                except:
                    pass
            seg_path = os.path.join(save_dir_class, f'{name}_seg.png')
            cam_path = os.path.join(save_dir_class, f'{name}_cam.png')
            cv2.imwrite(seg_path, heatmap)
            cv2.imwrite(cam_path, hm_on_img)
            print(f"[{_class_}] 保存结果: {seg_path} | Heatmap范围: {heatmap.min()}~{heatmap.max()}")
    return

**Comment by xsjcrl** [comment]
特别感谢您提供的代码，我相信它会对我有很大帮助。
